### PR TITLE
 Add ProtoStreamWriterInterface and implement forwarding writer.

### PR DESCRIPTION
**Pull request by cschuet**

### CODE REVIEW

File: cartographer/io/in_memory_proto_stream.h [hunk 1]
```diff
@@ -27,6 +27,25 @@
...
+class ForwardingProtoStreamWriter
+    : public cartographer::io::ProtoStreamWriterInterface {
+ public:
+  // A callback that is invoked anytime 'WriteProto()' is called on the
+  // 'ForwardingProtoStreamWriter'. When 'Close()' is called on the
+  // 'ForwardingProtoStreamWriter' the callback is invoked with a 'nullptr'.
+  using WriterCallback =
+      std::function<bool(const google::protobuf::Message* proto)>;
+
+  explicit ForwardingProtoStreamWriter(WriterCallback writer_callback);
```
Comment (pifon2a) [inline cartographer/io/in_memory_proto_stream.h]: Let's put constructor implementation to the header, like for InMemoryProtoStreamReader.

Comment (cschuet) [reply cartographer/io/in_memory_proto_stream.h]: Done.

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 1.

**Pull request merged**